File: driver/src/main/java/com/blundell/driver/Relay.java
package com.blundell.driver;

import com.google.android.things.pio.Gpio;
import com.google.android.things.pio.PeripheralManager;

import java.io.IOException;

public class Relay {

    public static Relay newInstance(String gpioPin) {
        PeripheralManager service = PeripheralManager.getInstance();
        return newInstance(service, gpioPin);
    }

    public static Relay newInstance(PeripheralManager service, String gpioPin) {
        try {
            Gpio gpio = service.openGpio(gpioPin);
            gpio.setActiveType(Gpio.ACTIVE_HIGH);
            gpio.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW);
            return new Relay(gpio);
        } catch (IOException e) {
            throw new IllegalStateException(e);
        }
    }

    private final Gpio gpio;

    Relay(Gpio gpio) {
        this.gpio = gpio;
    }

    public void connectTo(Position position) {
        try {
            switch (position) {
                case NORMALLY_OPEN:
                    gpio.setValue(true);
                    break;
                case NORMALLY_CLOSED:
                    gpio.setValue(false);
                    break;
                default:
                    throw new IllegalStateException("Developer Error, unhandled enum " + position);
            }
        } catch (IOException e) {
            throw new IllegalStateException("Cannot switch to " + position, e);
        }
    }

    public void close() {
        try {
            gpio.close();
        } catch (IOException e) {
            throw new IllegalStateException(e);
        }
    }

    public enum Position {
        NORMALLY_OPEN, NORMALLY_CLOSED
    }

}

File: app/src/main/java/com/blundell/driver/Relay.kt
package com.blundell.driver

import com.blundell.driver.RelayKotlin.PositionKotlin.NORMALLY_CLOSED
import com.blundell.driver.RelayKotlin.PositionKotlin.NORMALLY_OPEN
import com.google.android.things.pio.Gpio
import com.google.android.things.pio.PeripheralManagerService

class RelayKotlin(val gpio: Gpio) {

    companion object Factory {
        fun newInstance(gpioPin: String): RelayKotlin {
            return newInstance(PeripheralManagerService(), gpioPin)
        }

        fun newInstance(service: PeripheralManagerService, gpioPin: String): RelayKotlin {
            val gpio = service.openGpio(gpioPin)
            gpio.setActiveType(Gpio.ACTIVE_HIGH)
            gpio.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW)
            return RelayKotlin(gpio)
        }
    }

    fun connectTo(position: PositionKotlin) {
        when (position) {
            NORMALLY_OPEN -> gpio.setValue(true)
            NORMALLY_CLOSED -> gpio.setValue(false)
            else -> {
                throw IllegalStateException("Developer error, unhandled enum " + position)
            }
        }
    }

    fun close() {
        gpio.close()
    }

    enum class PositionKotlin {
        NORMALLY_OPEN, NORMALLY_CLOSED
    }

}
